### ISSUE TITLE
jQuery-基本

### ISSUE THREAD

**Issue by zjy7** [zjy7/blog#29]
.hide()
.show()
.slideDown()
.fadeIn()
.fadeOut()

$('#div1').slideDown(1000,function(){
     $('#div2').slideDown(2000);    //当div1完成后再展示div2
});

$('#div1').animate({
     width:"500px",
     marginLeft:"30px",
     borderWidth:"10px"
},1500);

**Comment by zjy7** [comment]
```
<!DOCTYPE html>
<html>
    <head>
        <meta charset="utf-8">
        <title>Challenge: Donut Calculator</title>
    </head>
    <body>

    <h1>Donut Calculator</h1>

    <p>Winston eats 3 donuts a day. How many donuts will he eat in his lifetime?</p>

    <form id="calculator-form">
        <label>
            Winston's final age:
            <input type="number" name="age">
        </label>
        <button type="submit">Calculate</button>
        <div id="calculator-results"></div>
    </form>

    <script src="https://ajax.googleapis.com/ajax/libs/jquery/2.1.4/jquery.min.js"></script>
    <script>
    var calcDonuts = function(years) {
        return years * 3 * 365;
    };

    $("#calculator-form").on("submit", function(event) {
        event.preventDefault();
        var v1 = $(this).find("[name=age]").val();
        //console.log(v1);
        var v2 = calcDonuts(v1);
        // TODO: Find inputted age, calculate # of donuts

        $("#calculator-results").text("Winston ate "+v2+ " donuts");
    });
    </script>
    </body>
</html>
```

**Comment by zjy7** [comment]
jQuery对象转换成DOM对象：
//第一种
var btn = $btn[0];
//第二种
var btn = $btn.get(0);

**Comment by zjy7** [comment]
jQuery版本
v1.11.3和v2.1.4是两个大版本的最新版

版本一：1.x版本
版本二：2.x版本

两个版本的区别：2.x版本不再支持IE678

**Comment by zjy7** [comment]
hide(时间)
show(时间)

slideUp
slideDown(1000)
slideToggle

.delay(1000)

fadeOut(1000)
fadeIn(1000)
fadeToggle

.toggle(1000)

.css({
     color:'red',
     fontWeight:'bold',
     display:'none'
})


.html('abc')
.html('ab<strong>my bold</strong>c')

$('#btn1').on('click',function(){
     //do something
});

事件
click
mouseover
mouseout


$('p#pid').html('para changed')

**Comment by zjy7** [comment]
<button class="panel-button" id="btn1" data-panel="panel1">BTN1</button>
<button class="panel-button" id="btn2" data-panel="panel2">BTN2</button>
<button class="panel-button" id="btn3" data-panel="panel3">BTN3</button>
<button class="panel-button" id="btn4" data-panel="panel4">BTN4</button>

$('button[data-panel]=panel1]')  //选中上面按钮

$('button[id=btn1]')     //选中上面按钮


//对所有panel-button类添加点击事件
$('.panel-button').on('click',function(){
     var panelId = $(this).attr('data-panel');
     //do something
})

**Comment by zjy7** [comment]
$('li')
$('li').first()
$('li').last()

$('li').eq(0)     //表示第一个li，还能显示它的行内css样式
$('li').eq(1)     //表示第二个li

$('ul:first').children() 

$('li:first').siblings()

$('li:first').parent()

.prev()     
.next()

.remove()

.cloest('   ')     //最近的

.find('  ')

.filter('   ')
 
if($(this).is('   ')){
      alert('aaa');
}

if($(this).not('   ')){

}

**Comment by zjy7** [comment]
<!DOCTYPE html>
<html>
    <head>
        <meta charset="utf-8">
        <title>Processing a quiz with jQuery</title>
    </head>
    <body>

    <h1>Reptile quiz</h1>

    <form id="quiz-form">
        <label>
            What is this? <br>
            <img src="https://ka-perseus-images.s3.amazonaws.com/325aab611b8b69b51bfeb16ade0af741da7dc3ff.jpg" alt="Reptile"> <br>
            <select id="quiz-answer">
                <option value="crocodile">Crocodile</option>
                <option value="alligator">Alligator</option>
                <option value="caiman">Caiman</option>
            </select>
        </label> <br>
        <button type="submit">Check Answer</button>
    </form>
    <div id="result"></div>

    <script src="https://ajax.googleapis.com/ajax/libs/jquery/2.1.4/jquery.min.js"></script>
    <script>

    // When the user submits the form,
    //   Check what answer they picked
    //   And tell them if they're correct

    // submit
    $("#quiz-form").on("submit", function(event) {
        event.preventDefault();                              //放置浏览器默认配置
        var $answer = $("#quiz-answer"); 
        var answer = $answer.val();                      //理解val()
        console.log(answer);
        if (answer === "crocodile") {
            $("#result").text("Woweeee! You got it! WOOO PARTY!");
        } else {
            $("#result").text("Try again!");
        }

    });
    </script>
    </body>
</html>

**Comment by zjy7** [comment]
**自定义事件**
var btn=$("#clickMeBtn");

btn.click(function(){
     var e = jQuery.Event("Myevent");
     btn.trigger(e);
});
btn.bind("Myevent",function(event){
     console.log(event);
});

**Comment by zjy7** [comment]
**禁用文本输入中的空格**
$('input.nospace').keydown(function(e){
    if(e.keyCode == 32){
        return false;
    }
}

**Comment by zjy7** [comment]
console.log("%c文案测试","font-size:50px; text-shadow:0 0 5px #f00;");

**Comment by zjy7** [comment]
jQuery选择的时候
ul li   
>  大于号
,   逗号
li:first
li:last
li:odd
li:even

$("ul").find("li")
$("li").first()
$("li").last()
$("li").first().next()
$("li").first().next().prev()
$("li").first().parent()
$("ul").children("li")

**Comment by zjy7** [comment]
.append(<element>)
.prepend(<element>)
.after(<element>)
.before(<element>)

var price = $("<p>This is a paragraph</p>");

.remove(<element>)


.appendTo(<element>)
.prependTo(<element>)
.insertAfter(<element>)
.insertBefore(<element>)

**Comment by zjy7** [comment]
$(document).ready(<event handler function>);

$('button').on('click',function(){

});

**Comment by zjy7** [comment]
$(this).closest('.div1').append(price);

**Comment by zjy7** [comment]
<div>
     <button>a</button>
</div>
$('button').on('click',function(){     });
$('div').on('click','button',function(){     });

**Comment by zjy7** [comment]
Mouse Events
click
dblclick
focusin
focusout
mousedown
mouseup
mousemove
mouseout
mouseover
mouseleave
mouseenter

**Comment by zjy7** [comment]
CSS Objects
.css(<attr>,<value>)
.css(<attr>)
.css(<object>)

$(this).css({
'background-color':'#ccc',
'border':'1px solid #F00'
});

**Comment by zjy7** [comment]
jquery 对外暴露了两个方法：jQuery和$
这两个方法实际上是同一个方法，
通过调用这两个方法，可以得到一个jQuery实例对象

jQuery实例对象是一个伪数组对象

jQuery和$实际上是一个工厂函数

**Comment by zjy7** [comment]
.slideDown()
.slideUp()
.slideToggle()

**Comment by zjy7** [comment]
button点击事件，尽量不要使用jquery中的事件委托代理delegate，
否则可能会产生意想不到的后果。
另外，该方法在jquery的高版本中已经被废除。


还有，搞清楚事件添加时，this的指向，尤其是异步操作时this可能会改变。
